client/network/src/protocol/generic_proto/behaviour.rs

```diff
@@ -1127,227 +1450,503 @@ impl NetworkBehaviour for GenericProto {
+					return
+				};
+
+				match mem::replace(entry.get_mut(), PeerState::Poisoned) {
+					// Enabled => Enabled | Disabled
+					PeerState::Enabled { mut connections } => {
+						debug_assert!(connections.iter().any(|(_, s)|
+							matches!(s, ConnectionState::Opening | ConnectionState::Open(_))));
+
+						let pos = if let Some(pos) = connections.iter().position(|(c, _)| *c == connection) {
+							pos
+						} else {
+							error!(target: "sub-libp2p",
+								"CloseDesired: State mismatch in the custom protos handler");
+							debug_assert!(false);
+							return;
 						};
 
-						(last, new_notifications_sink)
+						if matches!(connections[pos].1, ConnectionState::Closing) {
+							return;
+						}
+
+						debug_assert!(matches!(connections[pos].1, ConnectionState::Open(_)));
+						connections[pos].1 = ConnectionState::Closing;
+
+						debug!(target: "sub-libp2p", "Handler({}, {:?}) <= Close", source, connection);
+						self.events.push_back(NetworkBehaviourAction::NotifyHandler {
+							peer_id: source.clone(),
+							handler: NotifyHandler::One(connection),
+							event: NotifsHandlerIn::Close,
+						});
+
+						if let Some((replacement_pos, replacement_sink)) = connections
+							.iter()
+							.enumerate()
+							.filter_map(|(num, (_, s))| {
+								match s {
+									ConnectionState::Open(s) => Some((num, s.clone())),
+									_ => None
+								}
+							})
+							.next()
+						{
+							if pos <= replacement_pos {
+								debug!(target: "sub-libp2p", "External API <= Sink replaced({:?})", source);
+								let event = GenericProtoOut::CustomProtocolReplaced {
+									peer_id: source,
+									notifications_sink: replacement_sink,
+								};
+								self.events.push_back(NetworkBehaviourAction::GenerateEvent(event));
+								*entry.into_mut() = PeerState::Enabled { connections, };
+							}
+
+						} else {
+							// List of open connections wasn't empty before but now it is.
+							if !connections.iter().any(|(_, s)| matches!(s, ConnectionState::Opening)) {
+								debug!(target: "sub-libp2p", "PSM <= Dropped({:?})", source);
+								self.peerset.dropped(source.clone());
+								*entry.into_mut() = PeerState::Disabled {
+									connections, backoff_until: None
+								};
+							} else {
+								*entry.into_mut() = PeerState::Enabled { connections };
+							}
+
+							debug!(target: "sub-libp2p", "External API <= Closed({:?})", source);
+							let event = GenericProtoOut::CustomProtocolClosed {
+								peer_id: source,
+							};
+							self.events.push_back(NetworkBehaviourAction::GenerateEvent(event));
+						}
+					},
+
+					// All connections in `Disabled` and `DisabledPendingEnable` have been sent a
+					// `Close` message already, and as such ignore any `CloseDesired` message.
+					state @ PeerState::Disabled { .. } |
+					state @ PeerState::DisabledPendingEnable { .. } => {
+						*entry.into_mut() = state;
+						return;
 					},
 					state => {
```

Comment:
Do I understand correctly that it is not possible to receive a `NotifsHandlerOut::CloseDesired` for a `PeerState::Incoming`? Would it not be possible for a remote node to open and then close a connection before the local PSM has time to react?

Reply:
In `PeerState::Incoming`, the handler is expected to be in the closed state, and as such receiving a `CloseDesired` would be a bug.